plat/rockchip/rk3368/drivers/pmu/pmu.c

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2015, Fuzhou Rockchip Electronics Co., Ltd
+ * Author: Tony.Xie
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <arch_helpers.h>
+#include <assert.h>
+#include <debug.h>
+#include <delay_timer.h>
+#include <mmio.h>
+#include <platform_def.h>
+#include <pmu.h>
+#include <rk3368_def.h>
+#include <soc.h>
+
+/*
+ * The struct is used in pmu_cpus_on.S which
+ * gets the data of the struct by the following index
+ * #define PSRAM_DT_SYS_MODE	0x0
+ * #define PSRAM_DT_MPIDR	0x4
+ * #define PSRAM_DT_SP		0x8
+ * #define PSRAM_DT_DDR_FUNC	0x10
+ * #define PSRAM_DT_DDR_DATA	0x14
+ * #define PSRAM_DT_DDRFLAG	0x18
+ */
+typedef struct psram_data {
+	uint32_t sys_mode;
+	uint32_t boot_mpidr;
+	uint64_t sp;
+	uint32_t ddr_func;
+	uint32_t ddr_data;
+	uint32_t ddr_flag;
+} psram_data_t;
+
+static psram_data_t *psram_sleep_cfg = (psram_data_t *) PSRAM_DT_BASE;
+
+/*****************************************************************************
+ * sram only surpport 32-bits access
+ ******************************************************************************/
+static void
+u32_align_cpy(unsigned int *dst, const unsigned int *scr, size_t len)
+{
+	uint32_t i;
+
+	for (i = 0; i < len; i++)
+		dst[i] = scr[i];
+}
+
+void pmusram_prepare(void)
+{
+	uint32_t *sram_dst, *sram_src;
+	size_t sram_size = 2;
+
+	/*
+	 * pmu sram code and data prepare
+	 */
+	sram_dst = (uint32_t *)PMUSRAM_BASE;
+	sram_src = (uint32_t *)&pmu_cpuson_entrypoint_start;
+	sram_size = (uint32_t *)&pmu_cpuson_entrypoint_end -
+		    (uint32_t *)sram_src;
+
+	u32_align_cpy(sram_dst, sram_src, sram_size);
+
+	psram_sleep_cfg->sp = PSRAM_DT_BASE - 16;
```

Comment:
Apart from the security issue, this looks very fragile. What does this 16 offset represent?